[GLIB] Invalid write reported by Valgrind on GSocketMonitor::sourceSocketCallback

lauromoura opened the pull request:
#### 5d047d3c9925bc44725c9239c27f408c98929132
<pre>
[GLIB] Invalid write reported by Valgrind on GSocketMonitor::sourceSocketCallback
<a href="https://bugs.webkit.org/show_bug.cgi?id=283053">https://bugs.webkit.org/show_bug.cgi?id=283053</a>

Reviewed by Carlos Garcia Campos.

Move the callback cleanup to be the last operation on the monitor as
losing the callback might invoke the monitor&apos;s destructor right away.

Combined changes:
* Source/WTF/wtf/glib/GSocketMonitor.cpp:
(WTF::GSocketMonitor::socketSourceCallback):

Canonical link: <a href="https://commits.webkit.org/286596@main">https://commits.webkit.org/286596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24337b246f919fbacd8fb0e5a2583d7a0ca00d91

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55478 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80968 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27719 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64620 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3771 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59935 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18051 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49849 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65644 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40264 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47244 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23133 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26042 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69626 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68370 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23465 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82416 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75720 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3819 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2505 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68217 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3972 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65613 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67460 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16827 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11431 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9528 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97974 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3766 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21437 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3789 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7219 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5547 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/glib/GSocketMonitor.cpp:
```diff
@@ -47,8 +47,8 @@ gboolean GSocketMonitor::socketSourceCallback(GSocket*, GIOCondition condition,
 
     if (monitor->m_shouldDestroyCallback) {
-        monitor->m_callback = nullptr;
         monitor->m_shouldDestroyCallback = false;
+        monitor->m_callback = nullptr;
```
I think your solution is the best we can do without making GSocketMonitor refcounted. But please add a one liner comment to make sure we don't reintroduce this bug!

lauromoura committed:
[GLIB] Invalid write reported by Valgrind on GSocketMonitor::sourceSocketCallback
https://bugs.webkit.org/show_bug.cgi?id=283053

Reviewed by Carlos Garcia Campos.

Move the callback cleanup to be the last operation on the monitor as
losing the callback might invoke the monitor's destructor right away.

Combined changes:
* Source/WTF/wtf/glib/GSocketMonitor.cpp:
(WTF::GSocketMonitor::socketSourceCallback):

Canonical link: https://commits.webkit.org/286596@main
